src/utils/lock.ts

```diff
@@ -0,0 +1,100 @@
+import { ApplicationError } from "../dataEntities";
+
+/**
+ * Simple lock class.
+ *
+ * Use in a try/finally block, calling `release()` in the `finally` block to guarantee the lock is released in any circumstance.
+ */
+export class Lock {
+    private waiters: Array<() => void> = [];
+    private mLocked = false;
+
+    public get locked() {
+        return this.mLocked;
+    }
+
+    /**
+     * Acquires the lock; returns a promise that resolves when the lock is acquired.
+     */
+    public async acquire(): Promise<void> {
+        if (!this.mLocked) {
+            this.mLocked = true;
+            return Promise.resolve();
+        } else {
+            return new Promise(resolve => {
+                this.waiters.push(resolve);
+            });
+        }
+    }
+
+    /**
+     * Released the lock.
+     *
+     * @throws ApplicationError if the called when the lock was
+     */
+    public release() {
+        if (!this.mLocked) throw new ApplicationError("Tried to release a Lock that was not locked");
+
+        if (this.waiters.length > 0) {
+            // resolve the first waiter in the queue
+            const resolve = this.waiters.shift()!;
```

Comment:
Why does this need the assertion !

Reply:
Because shift() can return undefined when length === 0, and the type system can only infer non-null checks.